OneCommunityGlobal/HighestGoodNetworkApp
# Raul - New Timer System

# Description


PLEASE, DO NOT SUGGEST NEW FEATURES AND CHANGES, JUST TEST AND VERIFY IF IT'S WORKING. SUGGEST ONLY FIXES!

Please tag @rauleffting , when you find something, this user is not currently on the project.

This PR implements the New Timer system that can now synchronize between multiples tabs, detect whenever the user is inactive and adds a new way to track the time with a Countdown mode, while improving speed of the connection and reducing the load to MongoDB.

## Related PRS (if any):

This PR is related to this PR in [backend](https://github.com/OneCommunityGlobal/HGNRest/pull/271)

## Mainly changes explained:

* Add react-use-websocket and progressbar dependencies
* Update timer-service endpoint 
* Update Header to use new timer 
* Add NewTimer, Countdown, Stopwatch, SwitchTimer and TimerStatus Components


## How to test:

Make sure you have the most updated branch,if you have some conflict no worries, don't need to solve ! Just do "git checkout -f development && git branch -D  said-feat-realtime-clock && git fetch --prune" (This command will delete the current branch of you machine and pull the branch again fully updated)

* Checkout to this branch in both frontend and backend
* Run `npm install` on the frontend to install new dependencies and `npm start`
* Run `npm install` on the backend to install new dependencies, `npm build` to build the server and `npm start` to start the server
* Log as any user
* Verify new timer running on the left of Navbar
* Run the stopwatch or countdown
* Open a new tab and sees that the timer is running with the same amount of time elapsed
* It works even on anonymous tabs, but without sync.

## Screenshots or videos of changes:
[Timer Showdown Video](https://www.dropbox.com/s/ybwo1z07flkk0zt/lastPreview.mp4?dl=0)
[User Stop Condition Video](https://www.dropbox.com/s/fccti51ftj5x01y/userStopsCondition.mp4?dl=0)
![image](https://user-images.githubusercontent.com/33133120/212488784-573e4ee3-2605-48f1-b020-e513cd9d98c1.png)
![image](https://user-images.githubusercontent.com/33133120/212488798-19f944b3-39f4-4f59-a349-c76429aae005.png)

## Note:
This is a draft PR and any optimizations coming from the Seniors are welcome, this is one of the core feature of the HGN.
